GatewayPool fixes

marcvelmer opened the pull request:
Fixed two little bugs in the GatewayPool: Avoiding useless init call after discovery and fixed shifting for not repeating the same initial Gateway

marcvelmer committed:
Fixed useless GatewayPool init call after GatewayDiscovery

marcvelmer committed:
Fixed shifting in GatewayPool for not repeating same initial Gateway

emmdim commented on src/net/gateway-pool.ts:
```diff
@@ -39,10 +39,7 @@ export class GatewayPool {
-
-                return pool.init()
-                    .then(() => pool)
+                return new GatewayPool(bestNodes, params)
```
@marcvelmer  Why do you avoid the init inside the discover function?

marcvelmer replied:
Because the init is calling the init function of the active gateway, which ends up calling the init function of the dvote gateway. There the function isUp() is called, which is already being called before, in the run function of the discovery.
I found this by debugging the API calls made by the `getInfo()` function of the dvote gateway and thought that maybe we could save one API call by avoiding calling the init function of the pool at this point.

emmdim replied:
@marcvelmer  Nice catch :-)  I agree we should remove the duplicate isUp call, but I think though what we should modify is what is done during the init function

emmdim replied:
@marcvelmer maybe we could use the isReady function:
https://github.com/vocdoni/dvote-js/blob/4ff00fac504233fbee87e78641ba4f923d6677e8/src/net/gateway.ts#L135
And if isReady returns false then do a gateway refresh in the catch

marcvelmer replied:
I changed the behaviour of what is done inside the `init()` function in order to not duplicate a `isUp()` call, but I still need some review.

marcvelmer replied:
This will be reviewed again when refactoring the pool.

marcvelmer committed:
Added performance time metric to DVOTE and WEB3 GWs
Changed order in WEB3 initial calls for bailing earlier
Changed getBlockNums for 'lazy' or 'on demand' usage, only called when required
Added custom errors
Changed GWs sort criteria
Added new TODOs for `initEns` and `isReady` functions

emmdim commented on src/net/gateway-web3.ts:
```diff
@@ -252,10 +253,24 @@ export class Web3Gateway {
+        if (this.lastBlockNumber) {
+            return Promise.resolve()
+        } else if (!this._provider) {
+            return Promise.resolve()
```
@marcvelmer  why this is `promise.resolve()`?How to you actually know if the function failed? It should never throw?

marcvelmer replied:
That's a good question where I also had several doubts.

When I call `Promise.all()` what I really want to do is calling `Promise.allSettled()`, because I want `lastBlockNumber` value for all GWs in that Promise, no matter if one or several requests fail. The problem is that calling `Promise.reject()` stops the `Promise.all()`, that's why it should never throw.

Because I'm using `Promise.all()` instead of `Promise.allSettled()`, I have found a 'tricky' way of achieving this:

`await Promise.all(web3getBlockNumberCandidates.map(p => p.catch(e => e))).then()`

But I'm not sure this correct and/or an elegant way of solving this... But now we can reject the Promise and still proceed with all lastBlockNumber calls.

marcvelmer replied:
Ok, thanks to @brickpop I'm using TS 4.3.2 and I can use `Promise.allSettled()` which is the correct way of implementing what we wanted and having the corresponding `Promise.resolve()` or `Promise.reject()` in every situation.

marcvelmer replied:
This issue it's outdated because the code has changed. Anyway, there is an issue still pending with moving the TS version to 4.3.2

emmdim commented on src/net/gateway-web3.ts:
```diff
@@ -252,10 +253,24 @@ export class Web3Gateway {
+            return Promise.resolve()
+        } else if (!(this._provider instanceof JsonRpcProvider) && !(this._provider instanceof Web3Provider) &&
+            !(this._provider instanceof IpcProvider) && !(this._provider instanceof InfuraProvider)) {
+            return Promise.resolve()
```
@marcvelmer  same as above

marcvelmer replied:
Same problem as above

emmdim commented on src/net/gateway-discovery.ts:
```diff
@@ -1,24 +1,16 @@
-const MIN_ROUND_SUCCESS_COUNT = 2
-
-export type IGatewayDiscoveryParameters = {
+export interface IGatewayDiscoveryParameters {
```
@marcvelmer  FYI this is possibly breaking, we should add it on the changelog file

brickpop replied:
Changing from `type` to `interface` should not have any direct impact.
But +1 on the changelog

marcvelmer committed:
Modified `isReady()` functions in both DVote and Web3 GWs to be consistent with actual HTTP calls instead of Websocket
Added new `isPrepared()` function replacing `isReady()` internal usage
Added new TODO in `isPrepared()` for simplifying internal status calls

marcvelmer committed:
Fixed `getBlockNumber()`  in Web3 GW for rejecting Promise when failing

brickpop committed:
Typescript version upgrade

marcvelmer committed:
Fixed tricky way of doing all `Promises.all` with correct function `Promises.allSettled`

marcvelmer committed:
Changed `isUp()` for `check()` in Web3Gateway, returning a Promise.all() of all metrics to be set in each Gateway, achieving a better response time because of doing the old `isUp` in a parallel `Promise.all()`
Added some TODOs
Changed `tsconfig.json` target `es2020` for 'es6', but needs to be reviewed

marcvelmer committed:
Changed Web3 sort method for block numbers
Added few conditions for discarding Web3 Gateways
Partly refactored GatewayDiscovery class
Added few TODOs

marcvelmer committed:
Refactored sorting in a method and create node pairs compressed in a single function that will be removed in the future

marcvelmer committed:
Removed useless await/async

marcvelmer committed:
Fixed testing condition

marcvelmer committed:
Removed calculated `weight` metric from DVote Gateway
Changed DVote Gateways order criteria, now performance time is the preferred metric
Refactor some internal DVote Gateway functions and add some docs
Added some TODO related to custom error handling

marcvelmer committed:
Added `weight` property getter in DVote Gateway for backward compatibility with TODO for removing in the future

marcvelmer committed:
Refactor several functions of both DVote and Web3 Gateways
Refactor and fixed timeout handling when filtering healthy nodes
Added few TODOs and fixed minor things
Moved to target ES2020 for using `Promise.allSettled()`

marcvelmer committed:
Moved to target ES6 temporary

marcvelmer committed:
Removed `perf_hooks` and changed for `Date` in order to measure response time

emmdim committed:
Adds user parameters to GatewayDiscovery class and
 new ens parameter to control ENS requests during el discovery
 (default value false, non-breaking)

brickpop commented on src/net/gateway-discovery.ts:
```diff
@@ -165,75 +172,78 @@ export class GatewayDiscovery {
-                    web3Gateways = web3Gateways.concat(result.web3)
+            // The gateways to check in this round are filtered here
+            // We take those unchecked or those who timed out the round before
+            const dvoteNodes = discoveredDvoteNodes.slice().filter(
```
When doing a `filter` there is no need for cloning the array (slice)

brickpop commented on src/net/gateway-discovery.ts:
```diff
@@ -165,75 +172,78 @@ export class GatewayDiscovery {
+                        }
+                    })
+                } else {
+                    isDiscoveryRoundFinished = true
```
Why not simply a `break`  and a `while(true)` below?

marcvelmer replied:
What do you think about this?

```ts
// Launch the checking process for this round split by the maximum parallel requests defined
do {
    testDvote = dvoteGateways.length < minNumberOfGateways ? dvoteNodes.splice(0, this.PARALLEL_GATEWAY_TESTS) : []
    testWeb3 = web3Gateways.length < minNumberOfGateways ? web3Nodes.splice(0, this.PARALLEL_GATEWAY_TESTS) : []

    // Until there are gateways to test, the check round is not finished
    await this.selectActiveNodes(testDvote, testWeb3, timeout).then((result: IGatewayActiveNodes) => {
        dvoteGateways = dvoteGateways.concat(result.dvote)
        web3Gateways = web3Gateways.concat(result.web3)
    })
// Loop until the discovery round is finished
} while ((testDvote.length !== 0 || testWeb3.length !== 0) && (dvoteGateways.length < minNumberOfGateways || web3Gateways.length < minNumberOfGateways))
```

marcvelmer committed:
Fixed minor bugs and refactored the main loop in the discovery process

brickpop commented on src/net/gateway-discovery.ts:
```diff
@@ -1,252 +1,343 @@
+        this.bootnodesContentUri = params.bootnodesContentUri
+        this.minNumberOfGateways = params.numberOfGateways || this.MIN_NUMBER_GATEWAYS
+        this.timeout = params.timeout || GATEWAY_SELECTION_TIMEOUT
+        this.ens = params.ens || false
```
A parameter named `ens` feels ambiguous. I would use something like `resolveEns` or `resolveEnsDomains` for clarity. Both in the interface and the class variable.

brickpop commented on src/net/gateway-discovery.ts:
```diff
@@ -1,252 +1,343 @@
+     */
+    private static getGatewaysFromBootnodeData(): Promise<IGatewayActiveNodes> {
+        const networkId = this.networkId
+        return promiseWithTimeout(
```
```suggestion
        const prom = GatewayBootnode.getGatewaysFromUri(this.bootnodesContentUri)
          .catch(() => {
                throw new GatewayDiscoveryError(GatewayDiscoveryError.BOOTNODE_FETCH_ERROR)
          })
        return promiseWithTimeout(prom, this.timeout, GatewayDiscoveryError.BOOTNODE_TIMEOUT_ERROR)
```

brickpop commented on src/net/gateway-discovery.ts:
```diff
@@ -1,252 +1,343 @@
+            }, {})
+        )
+        if (blockNumbersByFrequency.length !== healthyNodes.web3.length) {
+            mostFrequentBlockNumber = +blockNumbersByFrequency.reduce((a, v) => (v[1] >= a[1] ? v : a), [null, 0])[0];
```
I think it more readable if the arguments of `reduce` are called something like `.reduce((prev, cur) => ... )`.

brickpop commented on src/net/gateway-dvote.ts:
```diff
@@ -98,46 +115,102 @@ export class DVoteGateway {
+            .then((request: MessageRequestContent) => {
+                requestId = request.id
+                return promiseWithTimeout(
+                    this.client.post('', JsonSignature.sort(request)),
```
This will probably have a conflict with existing code on `main`.

I removed `JsonSignature.sort()` from here and placed it when the request is created, before signing it. Now this code is in `createRequest` and we might need to port this and disregard the commit on `main`.

FYI: Internally `JsonSignature.sign()` also sorts the JSON fields so that signatures are reproducible, but currently gateways verify them using whatever JSON string they receive. So sorting, is really not needed anymore, just for consistency.

brickpop commented on src/net/gateway-dvote.ts:
```diff
@@ -98,46 +115,102 @@ export class DVoteGateway {
+                )
+            })
+            .then((response: AxiosResponse) => {
+                return this.checkRequest(response, requestId)
```
Shouldn't it be called like `checkResponse` ?

brickpop commented on src/net/gateway-dvote.ts:
```diff
@@ -98,46 +115,102 @@ export class DVoteGateway {
         }
-        const requestId = Random.getHex().substr(2, 10)
 
         const request: MessageRequestContent = {
```
The comment above about `JsonSignature.sign()` was meant to be applied here

brickpop commented on src/net/gateway-dvote.ts:
```diff
@@ -146,42 +219,34 @@ export class DVoteGateway {
-
-    /** Retrieves the status of the gateway and updates the internal status */
-    public updateGatewayStatus(timeout?: number): Promise<void> {
+    public check(timeout: number = GATEWAY_SELECTION_TIMEOUT): Promise<void> {
```
Given that `check` and `checkRequest => checkResponse` exist, it's not clear the difference between them. How about `checkStatus`?

brickpop commented on src/net/gateway-dvote.ts:
```diff
@@ -146,42 +219,34 @@ export class DVoteGateway {
-    /** Retrieves the status of the gateway and updates the internal status */
-    public updateGatewayStatus(timeout?: number): Promise<void> {
+    public check(timeout: number = GATEWAY_SELECTION_TIMEOUT): Promise<void> {
+        const performanceTime = new Date().getTime()
```
How about `checkPoint`? performancetime may be confusing

marcvelmer replied:
What about `responseTime`? The idea behind that is to store the time needed for the _last_ request done by the Gateway

brickpop commented on src/net/gateway-web3.ts:
```diff
@@ -190,69 +195,129 @@ export class Web3Gateway {
+     *
+     * @param timeout
+     */
+    public check(timeout: number = GATEWAY_SELECTION_TIMEOUT, ens: boolean = false): Promise<void> {
```
If `check`is renamed to `checkStatus` on the other GW, it would make sense to do the same here, and everywhere. (happy to hear feedback)

Same with `ens`, it may be confusing as a parameter.

brickpop commented on src/net/gateway-web3.ts:
```diff
@@ -190,69 +195,129 @@ export class Web3Gateway {
-        if (!this._provider) return Promise.resolve(0)
+    public getPeers(): Promise<void> {
+        // TODO Review if not rejecting and setting peerCount = -1 is the best solution
+        if (!this._provider) return Promise.reject()
```
May it be useful to return an Error, instead of nothing?

marcvelmer replied:
For the moment, the only Error thrown in the function is when the `peerCount` call exists, but fails. Otherwise we set the value of the `peerCount`

brickpop commented on src/net/gateway-web3.ts:
```diff
@@ -190,69 +195,129 @@ export class Web3Gateway {
         else if (!(this._provider instanceof JsonRpcProvider) && !(this._provider instanceof Web3Provider) &&
             !(this._provider instanceof IpcProvider) && !(this._provider instanceof InfuraProvider)) {
-            return Promise.resolve(0)
+            return Promise.reject()
```
Some gateways may not support this call, and returning `0`was a way of bypassing this in case of error. Like this, it is going to fail, instead of trying to using it anyway.

marcvelmer replied:
There are two changes in this function. The first one is about making the metric fetch "standard" as a `Promise<void>` and the second one is to set the `peerCount` to -1 if the web3 Gateway doesn't support the call, avoiding the usage of `0` as an error and using it as a real number. This is important when selecting the healthy nodes, because if `peerCount = -1` then there is no restriction for rejecting the Gateway, as we know that the call was not supported and therefore we cannot decide based on the number of peers.

brickpop commented on src/net/gateway-web3.ts:
```diff
@@ -190,69 +195,129 @@ export class Web3Gateway {
+    }
+
+    /** Request the block number of the Gateway which is currently connected to */
+    public getBlockNumber(): Promise<void> {
```
Why do we need a wrapper around `getBlockNumber`?

Do we simply want it to do the performance measurements?

marcvelmer replied:
There are two reasons for that. The first one is what you comment, the performance measurements, and the other one is to have some kind of "standard" inside the Web3Gateway, where every metric is a `Promise<void>` and the value of the metric (the block number in this case) is stored in a variable in the Object.

brickpop commented:
Also @marcvelmer let's consider https://www.npmjs.com/package/promise.allsettled as well, to see if we can leave the JS target as it was

marcvelmer committed:
Removed calculated `weight` metric from DVote Gateway
Changed DVote Gateways order criteria, now performance time is the preferred metric
Refactor some internal DVote Gateway functions and add some docs
Added some TODO related to custom error handling
